### PR TITLE
grpc server [2]: impl grpc server handle.

### PR DESCRIPTION
**Pull request by disksing**
The plan:
1. [x] update kvproto.
2. [x] remove kv/coprocessor from TCP server.
3. [x] impl  grpc server.
4. [ ] impl raft grpc client.
5. [ ] impl grpc snapshot.
6. [ ] replace TCP server with grpc servers.
7. [ ] add tests.

### CODE REVIEW

File: src/server/tikv_grpc.rs [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+use storage::{self, Storage, Key, Options, Mutation};
+use super::coprocessor::{RequestTask, EndPointTask, EndPointHost};
+use super::metrics::*;
+use super::Result;
+
+#[allow(dead_code)]
+const DEFAULT_COPROCESSOR_BATCH: usize = 50;
+
+#[allow(dead_code)]
+pub struct TiKVAsync {
```
Comment (siddontang) [inline src/server/tikv_grpc.rs]: AsyncServer? 
I don't think we need introduce name TiKV in our code.

File: src/server/tikv_grpc.rs [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+    let (tx, rx) = oneshot::channel();
+    let callback = move |resp| {
+        tx.send(resp).unwrap();
+    };
+    let future = rx.map_err(GrpcError::Canceled);
+    (box callback, box future)
+}
+
+impl tikvpb_grpc::TiKVAsync for TiKVAsync {
+    fn KvGet(&self, mut p: GetRequest) -> GrpcFutureSend<GetResponse> {
```
Comment (andelf) [inline src/server/tikv_grpc.rs]: never use CamelCase for fn names in hand written code, for rust.
😆

File: src/server/tikv_grpc.rs [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+                if let Err(e) = r {
+                    errs.push(extract_key_error(&e));
+                }
+            }
+        }
+        Err(e) => {
+            errs.push(extract_key_error(&e));
+        }
+    }
+    errs
```
Comment (andelf) [inline src/server/tikv_grpc.rs]: use into_iter() , map(), collect()?

File: src/server/tikv_grpc.rs (hunk 2 above)
Comment (disksing) [inline src/server/tikv_grpc.rs]: It's generated by grpc compiler.

File: src/server/handle.rs [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+                    }
+                };
+                future::result(res.map_err(|_| GrpcError::Other("schedule snap_task fail")))
+            })
+            .and_then(|_| future::ok::<_, GrpcError>(Done::new()))
+            .boxed()
+    }
+}
+
+use storage::txn::Error as TxnError;
```
Comment (siddontang) [inline src/server/handle.rs]: move the use to the top of the file

File: src/server/handle.rs [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+use storage::mvcc::Error as MvccError;
+use storage::engine::Error as EngineError;
+
+fn extract_region_error<T>(res: &storage::Result<T>) -> Option<RegionError> {
```
Comment (siddontang) [inline src/server/handle.rs]: do we have any test to cover these functions ?

Comment (disksing) [inline src/server/handle.rs]: These funcs are copied from https://github.com/pingcap/tikv/blob/master/src/server/kv.rs , there are some old tests, I'll take them back in later PR.

File: src/server/handle.rs [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+
+#[allow(dead_code)]
+pub struct Handle<T: RaftStoreRouter + 'static> {
+    // For handling KV requests.
+    storage: Mutex<Storage>,
+    // For handling coprocessor requests.
+    end_point_worker: Mutex<Worker<EndPointTask>>,
+    end_point_concurrency: usize,
+    // For handling raft messages.
+    ch: Mutex<T>,
```
Comment (siddontang) [inline src/server/handle.rs]: why using Mutex?

Comment (disksing) [inline src/server/handle.rs]: grpc server requires the Handle implements `Sync`.

Comment (siddontang) [inline src/server/handle.rs]: can we use RwLock, so we can use Read for concurrent access.

File: src/server/handle.rs [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+    let callback = move |resp| {
+        tx.send(resp).unwrap();
+    };
+    let future = rx.map_err(GrpcError::Canceled);
+    (box callback, box future)
+}
+
+impl<T: RaftStoreRouter + 'static> tikvpb_grpc::TiKVAsync for Handle<T> {
+    fn KvGet(&self, mut p: GetRequest) -> GrpcFutureSend<GetResponse> {
+        RECV_MSG_COUNTER.with_label_values(&["kv"]).inc();
```
Comment (siddontang) [inline src/server/handle.rs]: can we use histogram here?

File: src/server/handle.rs [hunk 8]
```diff
@@ -0,0 +1,565 @@
...
+                                          self.end_point_concurrency);
+        try!(self.end_point_worker
+            .lock()
+            .unwrap()
+            .start_batch(end_point, DEFAULT_COPROCESSOR_BATCH));
+        Ok(())
+    }
+}
+
+fn make_callback<T: Debug + Send + 'static>() -> (Box<FnBox(T) + Send>, GrpcFutureSend<T>) {
```
Comment (overvenus) [inline src/server/handle.rs]: How about returning a `oneshot::Receiver`? We can always box it later.

File: src/server/handle.rs [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+        let token = Token(self.token.fetch_add(1, Ordering::SeqCst));
+        let sched = self.snap_scheduler.lock().unwrap().clone();
+        let sched2 = sched.clone();
+        s.for_each(move |mut chunk| {
+                if chunk.has_message() {
+                    sched.schedule(SnapTask::Register(token, chunk.take_message()))
+                        .map_err(|_| GrpcError::Other("schedule snap_task fail"))
+                } else if !chunk.get_data().is_empty() {
+                    // TODO: Remove PipeBuffer or take good use of it.
+                    let mut b = PipeBuffer::new(chunk.get_data().len());
```
Comment (overvenus) [inline src/server/handle.rs]: It may block the tokio core.

File: src/server/handle.rs [hunk 10]
```diff
@@ -0,0 +1,565 @@
...
+                let res = match res {
+                    Ok(_) => sched2.schedule(SnapTask::Close(token)),
+                    Err(e) => {
+                        error!("receive snapshot err: {}", e);
+                        sched2.schedule(SnapTask::Discard(token))
+                    }
+                };
+                future::result(res.map_err(|_| GrpcError::Other("schedule snap_task fail")))
+            })
+            .and_then(|_| future::ok::<_, GrpcError>(Done::new()))
```
Comment (overvenus) [inline src/server/handle.rs]: It seems that this closure can be put in the `then` above.

File: src/server/handle.rs [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+            })
+            .then(move |res| {
+                let res = match res {
+                    Ok(_) => sched2.schedule(SnapTask::Close(token)),
+                    Err(e) => {
+                        error!("receive snapshot err: {}", e);
+                        sched2.schedule(SnapTask::Discard(token))
+                    }
+                };
+                future::result(res.map_err(|_| GrpcError::Other("schedule snap_task fail")))
```
Comment (overvenus) [inline src/server/handle.rs]: No need `future::result`.

File: src/server/handle.rs [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+        }
+    }
+}
+
+fn extract_key_errors(res: storage::Result<Vec<storage::Result<()>>>) -> Vec<KeyError> {
+    match res {
+        Ok(res) => {
+            res.into_iter()
+                .filter(|x| x.is_err())
+                .map(|e| extract_key_error(&e.unwrap_err()))
```
Comment (overvenus) [inline src/server/handle.rs]: How about `filter_map` ?

File: src/server/handle.rs (hunk 9 above)
Comment (disksing) [inline src/server/handle.rs]: Why? It's just a memory alloc.

File: src/server/handle.rs (hunk 8 above)
Comment (disksing) [inline src/server/handle.rs]: Seems box inside can reduce duplicated code.

File: src/server/handle.rs (hunk 7 above)
Comment (disksing) [inline src/server/handle.rs]: No, we can't. It returns a Future immediately.

File: src/server/handle.rs [hunk 13]
```diff
@@ -0,0 +1,565 @@
...
+use storage::txn::Error as TxnError;
+use storage::mvcc::Error as MvccError;
+use storage::engine::Error as EngineError;
+use super::transport::RaftStoreRouter;
+use super::coprocessor::{RequestTask, EndPointTask, EndPointHost};
+use super::snap::Task as SnapTask;
+use super::metrics::*;
+use super::Result;
+
+#[allow(dead_code)]
```
Comment (disksing) [inline src/server/handle.rs]: They will be removed in phase4.

**Pull request closed without merging**